web/src/backend/jwks.ts

```diff
@@ -57,28 +72,71 @@ export const fetchActiveJWK = async (alg: string) => {
     },
   });
 
-  if (!data.jwks?.length) {
-    throw new Error("No valid JWKs were found.");
+  // JWK is still active, so return
+  if (data.jwks?.length) {
+    const { id, kms_id } = data.jwks[0];
+    return { kid: id, kms_id };
   }
 
-  const { id, private_jwk } = data.jwks[0];
-  return { kid: id, private_jwk };
+  // No active JWK found, rotate the existing keys
+  const jwk = await _rotateJWK(alg);
+  return { kid: jwk.id, kms_id: jwk.kms_id };
 };
 
 /**
  * Generates an asymmetric key pair in JWK format
  * @returns
  */
-export const generateJWK = async (
-  alg: string
-): Promise<{
-  privateJwk: jose.JWK;
-  publicJwk: jose.JWK;
-}> => {
-  const { publicKey, privateKey } = await jose.generateKeyPair(alg);
-
-  const privateJwk = await jose.exportJWK(privateKey);
-  const publicJwk = await jose.exportJWK(publicKey);
-
-  return { privateJwk, publicJwk };
+export const generateJWK = async (alg: string): Promise<CreateJWKResult> => {
+  const client = await getKMSClient();
+
+  if (client) {
+    const result = await createKMSKey(client, "RSA_2048"); // TODO: transform alg parameter?
+    if (result?.keyId && result?.publicKey) {
+      const publicJwk = createPublicKey(result.publicKey).export({
+        format: "jwk",
+      });
+
+      return { keyId: result.keyId, publicJwk };
+    } else {
+      throw new Error("Unable to create KMS key.");
+    }
+  } else {
+    throw new Error("KMS client not found.");
+  }
+};
+
+/**
+ * Rotates the given JWK key by generating a new KMS key, and dropping the old value
+ */
+const _rotateJWK = async (alg: string) => {
+  // Delete the old JWK before generating a new one
```

Comment:
Not security related, but do you also delete the key from KMS? I think the KMS keys are billed at $1/month until deleted.

Reply:
Planning to add a Lambda to delete expired keys in a separate PR